### PR TITLE
Job base classes MSI support

### PR DESCRIPTION
**Pull request by drewgillies**
This is @agr's latest work on the job base types supporting MSIs for storage account access. Tested via other changes as discussed offline.

### CODE REVIEW

**Comment by joelverhagen** [comment]
Should this go into a feature branch instead, so we can stabilize all the jobs from there?

File: src/NuGet.Jobs.Common/StorageAccountExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+            if (builder == null)
+            {
+                throw new ArgumentNullException(nameof(builder));
+            }
+            if (getConnectionString == null)
+            {
+                throw new ArgumentNullException(nameof(getConnectionString));
+            }
+
+            return builder.Register(c =>
```
Comment (joelverhagen) [inline src/NuGet.Jobs.Common/StorageAccountExtensions.cs:90]: I wonder, should this not be registered as a singleton?
The recommendation from Azure SDK is that they are singleton:
https://learn.microsoft.com/en-us/dotnet/azure/sdk/thread-safety#client-lifetime

Comment (joelverhagen) [reply src/NuGet.Jobs.Common/StorageAccountExtensions.cs:90]: I think this general question applies to all SDK migration PRs. In short, are we adhering to the singleton recommendation provided by the SDK?

File: src/NuGetGallery.Core/Extensions/ConnectionStringExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace NuGetGallery
+{
+    public static class ConnectionStringExtensions
+    {
+        public static Uri GetBlobEndpointFromConnectionString(string connectionString)
+        {
+            var tempClient = new BlobServiceClient(connectionString);
+            // if _storageConnectionString has SAS token, Uri will contain SAS signature, we need to strip it
+            var uriBuilder = new UriBuilder(tempClient.Uri);
```
Comment (joelverhagen) [inline src/NuGetGallery.Core/Extensions/ConnectionStringExtensions.cs:19]: You can just do `GetLeftPart(UriPartial.Path)` instead of using `UriBuilder`.

**Review by joelverhagen — CHANGES_REQUESTED** [review]
Left comments.

File: src/NuGet.Jobs.Common/StorageAccountExtensions.cs (hunk 1 above)
Comment (agr) [reply src/NuGet.Jobs.Common/StorageAccountExtensions.cs:90]: We can't use singletons when we fall back to using credentials in storage connection strings (those have to be refreshed periodically). We can special case it, but then we'd have it as a singleton in one case and transient in the other, not sure what consequences would be.

**Comment by drewgillies** [comment]
Abandoning this PR as there was some confusion about the nature of this branch. I'll open a new PR against `agr-stsdk`.

**Comment by drewgillies** [comment]
Replaced by https://github.com/NuGet/NuGetGallery/pull/10153, later commits coming in later work.

**Pull request closed without merging**